### DISCUSSION TITLE
Concrete Executor Roadmap

### DISCUSSION THREAD

**Discussion by SBOne-Kenobi** [UnitTestBot/UTBotJava#1290] [Ideas]
What can Concrete Executor do:
- Build concrete values and instances from models given by Engine
- Run user's methods with collecting statistics such as coverage
- Mock user's functions corresponding to given models
- Mock new calls in function's body 

There are some enhancements for Concrete Executor, that should be done in the nearest future.

### Non-deterministic behavior mocking #1193  

There are situations when we don't analyze a code before executing it:
- Fuzzing
- Execution in the middle of symbolic analyzing

So, in the code may appear some functions which behavior is not deterministic such as random, thus result of execution may be not deterministic too. To fix it we can memorize results of executed functions that we consider non-deterministic, and respect them in final results, for instance, by mocking.

### Optimize symbolic analyzing by executing context-free methods

There are functions that don't depend on external variables, we can call this functions context-free. So, for symbolic analyze we can get just result from context-free methods without executing them symbolically.

### Dynamic symbolic execution

Developing previous idea we can use concrete executing instead symbolic one on functions and instructions that depend on only variables with concrete values. To do this we should analyze memory usage and separate it on concrete and symbolic types.

**Comment by sergeypospelov** [comment]
### Problems:

- Phases for `UtExecutionInstrumentation`
- Move `UtExecutionInstrumentation` to `utbot-instrumentation` module
- Unclear API for `ConcreteExecutor::executeAsync`
- Managed `Settings`
- Configuration issues
- Useless `RuntimeTraceStorage` allocation in the main process
- Refactor tracing instrumentation
- Problem with classloaders
